### PR TITLE
Declare constraints and transforms for distribution parameters

### LINKED ISSUE #55: Register .transform methods for distribution parameters
PR https://github.com/pytorch/pytorch/pull/4902

Each distribution should know the support of its parameters and results (e.g. `(-inf,inf)`, `(0,inf)`, `(0,1)`). This issue is about registering transform pairs for parameters (unlike distribution outputs, parameters do not need a `.log_abs_det_jacobian()` method and so do not need a full `Bijector` class).

See [relevant section in the Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.unemmumtnzfk). See also #56

### PR DESCRIPTION
**Pull request by fritzo**
This adds a `.params` and a `.support` attribute to each distribution. The purpose is to declare the support set of each parameter of a distribution and samples drawn from the distribution. These declarations are used in the `Constraint` class hierarchy and for transforming to and from unconstrained spaces.

This PR registers pseudoinverse pairs (`.to_unconstrained()`, `.to_constrained()`) that are used to transform top-level distribution parameters during optimization. These functions may map between spaces of different dimensionality (e.g. for `simplex`) and therefore are not bijections.

This PR does not register `Bijector`s for transforming sample results for use in HMC. This more complex task will be addressed in #56 and requires an implementation of `Bijector` classes with `.log_abs_det_jacobian()` methods.

Until there's downstream machinery to make use of discrete constraints, I've simply declared all discrete variables as `unconstrained`. We can make these more specific in the future to support say Gibbs sampling, e.g. `nonnegative_integer` or `integer_interval(low, high)`.

See [Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.unemmumtnzfk)
Fixes #55

### CODE REVIEW

File: torch/distributions/constraints.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        raise ValueError('Parameter cannot be transformed from an unconstrained space; '
+                         'Try another parameterization to make parameters independent.')
+
+    def to_constrained(self, u):
+        raise ValueError('Parameter cannot be transformed to an unconstrained space; '
+                         'Try another parameterization to make parameters independent.')
+
+
+class Positive(Constraint):
+    """
```
Comment (neerajprad) [inline torch/distributions/constraints.py]: This is nice! Do we see a need to implement multiple transform pairs for each constraint type in the future? If so, each constraint type could register multiple transform pairs (or register bijectors to transform samples), with a default pair like the one you already have. We can probably defer this until we really need it though.

Comment (fritzo) [reply torch/distributions/constraints.py]: It seems like a reasonable pattern for extension is to subclass `Positive` with something like
```py
class PositiveWithSoftplus(Positive):
    def to_constrained(u): ...
    def to_unconstrained(x): ...
```
The other direction `Constraint` class will grow is to add `.bijector` properties which are different from the constraint transforms implemented in this PR. These will look something like
```py
class Positive(Constraint):
    def to_constrained(u): ...
    def to_unconstrained(x): ...
    bijector = [LogBijector()]

class Interval(Constraint):
    ...
    @property
    def bijector(self):
        return [AffineBijector(self.lower_bound, self.upper_bound - self.lower_bound),
                SigmoidBijector()]
```

Comment (neerajprad) [reply torch/distributions/constraints.py]: Nice, I like it.

**Comment by fritzo** [comment]
@alicanb @jwvdm: Neeraj and I would like this `.constraint` information in Pyro. Does this design look suitable also for ProbTorch? If so I'll send the PR to pytorch/pytorch.

File: torch/distributions/constraints.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import torch
+from torch.nn.functional import sigmoid, softmax
+
+
+class Constraint(object):
```
Comment (fritzo) [inline torch/distributions/constraints.py]: In a subsequent PR addressing #56 I plan to add `.bijector` attributes to this class. The `.to_*()` methods are aimed at SVI whereas the `.bijector` methods are aimed at HMC.

**Comment by jwvdm** [comment]
This looks great. I do have some high-level questions about the API design:

- Is the "sample" constraint just the support of the distribution (i.e. it formulates a constraint on values returned by sample, and inputs to log_prob)? If so, then could we simply call it "support"? 

- What is the generic API interface for getting untransformed parameters associated with an arbitrary distribution? In other words, suppose I want to get unconstrained values for all parameters of a distribution `d`, then will there be a method `d.parameters().unconstrained()`, or something along those lines, or do I need to know that `d` is of type, e.g. Normal, and that normal has parameters named `loc` and `scale`?

- Finally, which of these features should live in PyTorch? I agree that being able to transformed random variables and/or distribution parameters to unconstrained values is very useful. However, whether you need this or not depends on the inference strategy. Arguably, the actual transformation code could should live in Pyro, whereas the distributions library should simply implement properties that indicate what the support is for random variables, and what constraints are present for parameter values.

**Comment by fritzo** [comment]
@jwvdm Thanks for the feedback! I've renamed "sample" to "support" as you suggest, and I think I'll change that to a `.support` attribute so that support can depend on parameters.

> which of these features should live in PyTorch? ... transformation code could live in Pyro

That's a great point, and at very least we could more cleanly separate constraints from transforms. I'll sketch that change since it seems like a middle ground: PyTorch could provide example transforms and users can either import or copy the design patterns to implement their own.

> What is the generic API for getting untransformed parameters of a distribution?

Neeraj and I [sketched some possible interfaces](https://colab.research.google.com/notebook#fileId=18D5dG-PMK0thXqI9soGE0BkKOQdGBX3H) and it seemed like there's some tension between (a) wanting to implement interpretable Bayesian models with meaningful params and (b) implementing generic black box inference algorithms. This PR takes a middle ground by assuming users know at least the name of each param (it looked gross to lump all params into a single black-box `Variable`). Using this PR in Pyro we could
```py
params = {}
for name in ['loc', 'scale']:  # <--- user at least needs to know names of params.
    c = normal.constraints[name]
    params[name] = c.to_constrained(pyro.param('unconstrained_' + name, torch.zeros(1)))
x = pyro.sample('x', normal, **params)
```

**Comment by fritzo** [comment]
@jwvdm ok by splitting `.constraints` into `.params` and `.support` we can now generically inspect each distribution's set of parameters :smile: (users still need to supply tensor dimensions).
```py
dist = torch.distributions.Normal  # <--- any other distribution will also work
params = {
    name: c.to_constrained(pyro.param('unconstrained_' + name, Variable(torch.zeros(1))))
    for name, in normal.params.items()
}
pyro.sample('x', dist(**params))```

**Comment by fritzo** [comment]
Ok I've split `transforms` out of `constraints`. If `transforms` seems too high-level for PyTorch, we'll keep it in Pyro; if it seems more generally useful we can merge it into PyTorch.

File: torch/distributions/constraints.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+from torch.nn.functional import sigmoid, softmax
+from torch.distributions.utils import tril_mask
+
+
+class Constraint(object):
+    """
+    Abstract base class for constraints.
+
+    A constraint object represents a region over which a continuous variable is
+    valid, e.g. within which a variable can optimized.
```
[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/constraints.py]: Thanks, done.

**Comment by fritzo** [comment]
Thanks Neeraj, I'll split this into two PRs and move it upstream.

**Pull request closed without merging**